README
NAME
    Test::Reporter - sends test results to cpan-testers@perl.org

VERSION
    version 1.60

SYNOPSIS
      use Test::Reporter;

      my $reporter = Test::Reporter->new(
          transport => 'File',
          transport_args => [ '/tmp' ],
      );

      $reporter->grade('pass');
      $reporter->distribution('Mail-Freshmeat-1.20');
      $reporter->send() || die $reporter->errstr();

      # or

      my $reporter = Test::Reporter->new(
          transport => 'File',
          transport_args => [ '/tmp' ],
      );

      $reporter->grade('fail');
      $reporter->distribution('Mail-Freshmeat-1.20');
      $reporter->comments('output of a failed make test goes here...');
      $reporter->edit_comments(); # if you want to edit comments in an editor
      $reporter->send() || die $reporter->errstr();

      # or

      my $reporter = Test::Reporter->new(
          transport => 'File',
          transport_args => [ '/tmp' ],
          grade => 'fail',
          distribution => 'Mail-Freshmeat-1.20',
          from => 'whoever@wherever.net (Whoever Wherever)',
          comments => 'output of a failed make test goes here...',
          via => 'CPANPLUS X.Y.Z',
      );
      $reporter->send() || die $reporter->errstr();

DESCRIPTION
    Test::Reporter reports the test results of any given distribution to the
    CPAN Testers project. Test::Reporter has wide support for various
    perl5's and platforms.

    CPAN Testers no longer receives test reports by email, but reports still
    resemble an email message. This module has numerous legacy "features"
    left over from the days of email transport.

  Transport mechanism
    The choice of transport is set with the "transport" argument. CPAN
    Testers should usually install Test::Reporter::Transport::Metabase and
    use 'Metabase' as the "transport". See that module for necessary
    transport arguments. Advanced testers may wish to test on a machine
    different from the one used to send reports. Consult the CPAN Testers
    Wiki <http://wiki.cpantesters.org/> for examples using other transport
    classes.

    The legacy email-based transports have been split out into a separate
    Test::Reporter::Transport::Legacy distribution and methods solely
    related to email have been deprecated.

ATTRIBUTES
  Required attributes
    *   distribution

        Gets or sets the name of the distribution you're working on, for
        example Foo-Bar-0.01. There are no restrictions on what can be put
        here.

    *   from

        Gets or sets the e-mail address of the individual submitting the
        test report, i.e. "John Doe <jdoe@example.com>".

    *   grade

        Gets or sets the success or failure of the distributions's 'make
        test' result. This must be one of:

          grade     meaning
          -----     -------
          pass      all tests passed
          fail      one or more tests failed
          na        distribution will not work on this platform
          unknown   tests did not exist or could not be run

  Transport attributes
    *   transport

        Gets or sets the transport type. The transport type argument is
        refers to a 'Test::Reporter::Transport' subclass. The default is
        'Null', which uses the Test::Reporter::Transport::Null class and
        does nothing when "send" is called.

        You can add additional arguments after the transport selection.
        These will be passed to the constructor of the lower-level
        transport. See "transport_args".

         $reporter->transport(
             'File', '/tmp'
         );

        This is not designed to be an extensible platform upon which to
        build transport plugins. That functionality is planned for the
        next-generation release of Test::Reporter, which will reside in the
        CPAN::Testers namespace.

    *   transport_args

        Optional. Gets or sets transport arguments that will used in the
        constructor for the selected transport, as appropriate.

  Optional attributes
    *   comments

        Gets or sets the comments on the test report. This is most commonly
        used for distributions that did not pass a 'make test'.

    *   debug

        Gets or sets the value that will turn debugging on or off. Debug
        messages are sent to STDERR. 1 for on, 0 for off. Debugging
        generates very verbose output and is useful mainly for finding bugs
        in Test::Reporter itself.

    *   dir

        Defaults to the current working directory. This method specifies the
        directory that write() writes test report files to.

    *   timeout

        Gets or sets the timeout value for the submission of test reports.
        Default is 120 seconds.

    *   via

        Gets or sets the value that will be appended to X-Reported-Via,
        generally this is useful for distributions that use Test::Reporter
        to report test results. This would be something like "CPANPLUS
        0.036".

  Deprecated attributes
    CPAN Testers no longer uses email for submitting reports. These
    attributes are deprecated.

    *   address

    *   mail_send_args

    *   mx

METHODS
    *   new

        This constructor returns a Test::Reporter object.

    *   perl_version

        Returns a hashref containing _archname, _osvers, and _myconfig based
        upon the perl that you are using. Alternatively, you may supply a
        different perl (path to the binary) as an argument, in which case
        the supplied perl will be used as the basis of the above data.

    *   subject

        Returns the subject line of a report, i.e. "PASS Mail-Freshmeat-1.20
        Darwin 6.0". 'grade' and 'distribution' must first be specified
        before calling this method.

    *   report

        Returns the actual content of a report, i.e. "This distribution has
        been tested as part of the cpan-testers...". 'comments' must first
        be specified before calling this method, if you have comments to
        make and expect them to be included in the report.

    *   send

        Sends the test report to cpan-testers@perl.org via the defined
        "transport" mechanism. You must check errstr() on a send() in order
        to be guaranteed delivery.

    *   edit_comments

        Allows one to interactively edit the comments within a text editor.
        comments() doesn't have to be first specified, but it will work
        properly if it was. Accepts an optional hash of arguments:

        *   suffix

            Optional. Allows one to specify the suffix ("extension") of the
            temp file used by edit_comments. Defaults to '.txt'.

    *   errstr

        Returns an error message describing why something failed. You must
        check errstr() on a send() in order to be guaranteed delivery.

    *   write and read

        These methods are used in situations where you wish to save reports
        locally rather than transmitting them to CPAN Testers immediately.
        You use write() on the machine that you are testing from, transfer
        the written test reports from the testing machine to the sending
        machine, and use read() on the machine that you actually want to
        submit the reports from. write() will write a file in an internal
        format that contains 'From', 'Subject', and the content of the
        report. The filename will be represented as:
        grade.distribution.archname.osvers.seconds_since_epoch.pid.rpt.
        write() uses the value of dir() if it was specified, else the cwd.

        On the machine you are testing from:

          my $reporter = Test::Reporter->new
          (
            grade => 'pass',
            distribution => 'Test-Reporter-1.16',
          )->write();

        On the machine you are submitting from:

          # wrap in an opendir if you've a lot to submit
          my $reporter;
          $reporter = Test::Reporter->new()->read(
            'pass.Test-Reporter-1.16.i686-linux.2.2.16.1046685296.14961.rpt'
          )->send() || die $reporter->errstr();

        write() also accepts an optional filehandle argument:

          my $fh; open $fh, '>-';  # create a STDOUT filehandle object
          $reporter->write($fh);   # prints the report to STDOUT

  Deprecated methods
    *   message_id

CAVEATS
    If you experience a long delay sending reports with Test::Reporter, you
    may be experiencing a wait as Test::Reporter attempts to determine your
    email address. Always use the "from" parameter to set your email address
    explicitly.

SEE ALSO
    For more about CPAN Testers:

    *   CPAN Testers reports <http://www.cpantesters.org/>

    *   CPAN Testers wiki <http://wiki.cpantesters.org/>

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/cpan-testers/Test-Reporter/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/cpan-testers/Test-Reporter>

      git clone https://github.com/cpan-testers/Test-Reporter.git

AUTHORS
    *   Adam J. Foxson <afoxson@pobox.com>

    *   David Golden <dagolden@cpan.org>

    *   Kirrily "Skud" Robert <skud@cpan.org>

    *   Ricardo Signes <rjbs@cpan.org>

    *   Richard Soderberg <rsod@cpan.org>

    *   Kurt Starsinic <Kurt.Starsinic@isinet.com>

CONTRIBUTORS
    *   Andreas Koenig <andk@cpan.org>

    *   Tatsuhiko Miyagawa <miyagawa@bulknews.net>

    *   Vincent Pit <perl@profvince.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Authors and Contributors.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Test::Reporter.

For the complete changelog, visit:
http://github.com/dagolden/test-reporter/commits/master

1.60      2013-11-04 13:49:54 America/New_York

        [FIXED]

        - read() now fixes up the architecture name

        - fixed documentation example for transport arguments
          involving File transport

1.59      2013-06-05 23:10:30 America/New_York


        [FIXED]

        - Fixed bug in finding right perl for -V output
          [rt.cpan.org #85823]

1.58      2011-07-19 12:07:53 America/New_York

        - Deprecates email-transport related attributes and methods

        - Adds new 'Null' transport as default

        - Removes legacy transport classes
          (See Test::Reporter::Transport::Legacy if needed)

        - Removes 'cpantest' command line program

1.57      2010-03-26 00:13:18 EST5EDT

        - Includes full 'perl -V' output, not just Config::mychanges

        - Adds new 'X-Test-Reporter-Perl' header with Perl version 
          and optional "RCx" suffix to positively identify tests from
          RC Perls

        - Minimum perl version is now 5.006, matching the rest of the Perl toolchain

1.56      2010-03-14 23:26:00 EST5EDT

        - Convert line endings during read() [Issue identified
          by Serguei Trouchelle]

1.55      2010-03-12 11:44:05 EST5EDT

        - Preserve distfile through write() and read() 
          [patch by Dan Collins]

        - Regenerate _myconfig data during read() method

        - Switched to Dist::Zilla for release management

1.54 Sun May 31 21:39:12 EDT 2009

        - POD updates.

        - Update Module::Install to 0.91.

1.53_03 Thu Apr 16 12:30:49 EDT 2009

        - Transmit report body line-by-line with Net::SMTP (or Net::SMTP::TLS).

1.53_02 Wed Apr 15 13:30:53 EDT 2009

        - With Net::SMTP transport, if lines are longer than 100 characters,
          the email will be encoded with quoted-printable.

1.53_01 Tue Apr  7 11:01:37 EDT 2009

        - Added optional 'distfile' argument to support
          T::R::Transport::Metabase.

        - Reports link to the wiki rather than www.cpantesters.org; also added
          cpan-testers-discuss as place to send Q's.

        - Improved diagnostic output if editor fails to launch.

        - Changed Makefile.Pl to Module::Install.

1.5203 Tue Jan 20 15:25:07 EST 2009

        - Updates CPAN Testers urls to www.cpantesters.org (David Westbrook).

1.5202 Mon Nov  3 11:48:44 EST 2008

        - Fixed bug in Mail::Send transport (GoogleCode issue 19; patch by
          Moshe Kamensky).

1.5201 Tue Sep 23 11:18:34 EDT 2008

        - All tests now use Test::More (not Test) for better diagnostics.

        - Removed unused signature test.

1.52 Mon Sep  8 11:45:12 EDT 2008

        - Fixed extraneous-newlines bug in reading reports saved to a file
          (Sleven Rezick).

1.51_02 Sun Sep  7 23:24:47 EDT 2008

        - Cache mail domain; shorten mail domain timeout; add 'smtp' as a
          hostname to check (KANE).

1.51_01 Sat Sep  6 14:34:13 EDT 2008

        - Avoid exclusive lock on tempfile for editing report (Slaven Rezick).

1.50 Sat Sep  6 14:04:22 EDT 2008

        - Deprecated sending reports to anywhere other than the cpan-testers
          mailing list.  This means that authors will no longer be copied on
          report emails by individual testers. (CPAN Testers is switching to
          an opt-in philosophy with central author notifications).

        - Doc fix on usage of read() in Test::Reporter::Transport::File.

1.4002 Wed Jun 25 17:21:10 EDT 2008

        - Fixed bug in error handling when using TLS transport.

1.4001 Wed May 21 12:34:28 EDT 2008

        - Fixed bug where send() would crash when no explicit transport was
          set.

1.40 Sun May 18 14:54:46 EDT 2008

        - No change from 1.39_06.

1.39_06 Thu Apr 17 06:15:45 EDT 2008

        - Fixed bug setting transport_args in constructor.

1.39_05 Wed Apr  9 12:43:55 EDT 2008

        - Removed use of File::Temp->newdir() for backwards compatibility with
          older versions of File::Temp.

1.39_04 Mon Apr  7 17:25:35 EDT 2008

        - Added Test::Reporter::Transport::File subclass.

1.39_03 Mon Apr  7 10:00:00 EDT 2008

        - Added documentation for transport.

1.39_02 Sun Apr  6 14:40:00 EDT 2008

        - Refactored transport into Test::Reporter::Transport subclasses.

1.39_01 Sun Apr  6 01:16:10 EDT 2008

        - Whoops! Devel versions are supposed to be odd.

        - t/2-http-transport.t left out of MANIFEST.

        - Copyright notice updates.

1.38_01 Sat Apr  5 18:00:00 EDT 2008

        - Add support for 'HTTP' transport to a Test::Reporter::HTTPGateway
          server.

1.38 Tue Oct 09 21:00:00 EDT 2007

        - Additional compatibility improvements for VMS. Many thanks to Peter
          (Stig) Edwards!

1.36 Mon Oct 08 21:00:00 EDT 2007

        - Compat for perl-5.005_05 per David Golden and Slaven Rezic.

        - Per 29690@RT cleaning up cpantest's -dump output (thanks janus!).

        - Per 9@Google compat fixes for VMS courtesy of Peter (Stig) Edwards.

1.34 Sun Aug 12 23:00:00 EDT 2007

        - It is now possible to override the automatic selection of a transport
          by making use of the transport() method (per WYANT and ANDK).

        - A dozen new tests added to support transport().

        - Internal debugging improvements.

        - Various documentation improvements.

        - Removing subversion artifacts; This project is now under git control.

1.32 Sun Jul 29 06:30:00 EDT 2007

        - Fixed an issue (8@Google) whereby test reports had Date headers that
          were not RFC-822 complaint (reported by Serguei Trouchelle).

1.30 Wed Jul 18 04:50:00 EDT 2007

        - Public release of all activity against 1.29_0* (see below).

        - Additionally, updated some documentation and added more tests.

1.29_04 Wed Jun 27 13:20:00 CDT 2007


lib/Test/Reporter/Transport.pm
use strict;
BEGIN{ if (not $] < 5.006) { require warnings; warnings->import } }
package Test::Reporter::Transport;
our $VERSION = '1.60'; # VERSION

sub new     { die "Not implemented" }

sub send    { die "Not implemented" }

1;

# ABSTRACT: base class for Test::Reporter transports

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Reporter::Transport - base class for Test::Reporter transports

=head1 VERSION

version 1.60

=head1 SYNOPSIS

    # Defines API that must be implemented by subclasses

    my $sender = Test::Reporter::Transport::Subclass->new( @args );

    $sender->send( $report )

=head1 DESCRIPTION

Transport subclasses provide the means by which CPAN Testers reports are 
transmitted to the CPAN Testers mailing list.

This module is an abstract base class that define an API for
Test::Reporter::Transport subclasses.  Individual subclasses MUST 
implement the methods described below.

=head1 USAGE

A transport method is specified to Test::Reporter using the C<transport>
option.  The C<transport> option expects just the module "suffix" that follows
C<Test::Reporter::Transport>.  For example:

    # use Test::Reporter::Transport::Null for transport
    my $report = Test::Reporter->new(
        transport => 'Null'
    );

Configuration of the transport is specified with the C<transport_args>
option:

    my $report = Test::Reporter->new(
        transport => 'File',
        transport_args => [ dir => '/tmp' ],
    );

These may also be specified with the C<transport> or C<transport_args> methods:

    $report->transport_args( dir => '/tmp' );

These may also be combined in the C<transport> method itself:

    $report->transport( 'File', dir => '/tmp' );

=head1 METHODS

The terms 'may', 'must', 'should', etc. used below have their usual RFC
meanings.

=head2 new

    my $sender = $subclass->new( @args );

The C<new> method is the object constructor.  It MAY take a list of any
necessary configuration options.  It MUST return a transport object if one
is successfully created or undef if the object can not be created. 

=head2 send

    $sender->send( $report );

The C<send> method MUST take a Test::Reporter object as its only argument.  It
MUST return true if the report is successfully sent.  It SHOULD die with a
message describing the failure if a report cannot be sent.  It MUST NOT return
a true value if the report cannot be sent.

=head1 AUTHORS

=over 4

=item *

Adam J. Foxson <afoxson@pobox.com>

=item *

David Golden <dagolden@cpan.org>

=item *

Kirrily "Skud" Robert <skud@cpan.org>

=item *

Ricardo Signes <rjbs@cpan.org>

=item *

Richard Soderberg <rsod@cpan.org>

=item *

Kurt Starsinic <Kurt.Starsinic@isinet.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Authors and Contributors.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Test/Reporter/Transport/File.pm
use strict;
BEGIN{ if (not $] < 5.006) { require warnings; warnings->import } }
package Test::Reporter::Transport::File;
our $VERSION = '1.60'; # VERSION

use base 'Test::Reporter::Transport';

sub new {
  my ($class, $dir) = @_;

  die "target directory '$dir' doesn't exist or can't be written to"
    unless -d $dir && -w $dir;

  return bless { dir => $dir } => $class;
}

sub send {
    my ($self, $report) = @_;
    $report->dir( $self->{dir} );
    return $report->write();
}

1;

# ABSTRACT: File transport for Test::Reporter

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Reporter::Transport::File - File transport for Test::Reporter

=head1 VERSION

version 1.60

=head1 SYNOPSIS

    my $report = Test::Reporter->new(
        transport => 'File',
        transport_args => [ $dir ],
    );

=head1 DESCRIPTION

This module saves a Test::Reporter report to the specified directory (using
the C<write> method from Test::Reporter.

This lets you save reports during offline operation.  The files may later be
uploaded using C<< Test::Reporter->read() >>.

    Test::Reporter->new->read( $file )->send();

=head1 USAGE

See L<Test::Reporter> and L<Test::Reporter::Transport> for general usage
information.

=head2 Transport Arguments

    $report->transport_args( $dir );

This transport class must have a writeable directory as its argument.

=head1 METHODS

These methods are only for internal use by Test::Reporter.

=head2 new

    my $sender = Test::Reporter::Transport::File->new( $dir ); 

The C<new> method is the object constructor.   

=head2 send

    $sender->send( $report );

The C<send> method transmits the report.  

=head1 AUTHORS

=over 4

=item *

Adam J. Foxson <afoxson@pobox.com>

=item *

David Golden <dagolden@cpan.org>

=item *

Kirrily "Skud" Robert <skud@cpan.org>

=item *

Ricardo Signes <rjbs@cpan.org>

=item *

Richard Soderberg <rsod@cpan.org>

=item *

Kurt Starsinic <Kurt.Starsinic@isinet.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Authors and Contributors.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


inc/Module/Install/Makefile.pm
#line 1
package Module::Install::Makefile;

use strict 'vars';
use Module::Install::Base;
use ExtUtils::MakeMaker ();

use vars qw{$VERSION $ISCORE @ISA};
BEGIN {
	$VERSION = '0.81';
	$ISCORE  = 1;
	@ISA     = qw{Module::Install::Base};
}

sub Makefile { $_[0] }

my %seen = ();

sub prompt {
	shift;

	# Infinite loop protection
	my @c = caller();
	if ( ++$seen{"$c[1]|$c[2]|$_[0]"} > 3 ) {
		die "Caught an potential prompt infinite loop ($c[1]|$c[2]|$_[0])";
	}

	# In automated testing, always use defaults
	if ( $ENV{AUTOMATED_TESTING} and ! $ENV{PERL_MM_USE_DEFAULT} ) {
		local $ENV{PERL_MM_USE_DEFAULT} = 1;
		goto &ExtUtils::MakeMaker::prompt;
	} else {
		goto &ExtUtils::MakeMaker::prompt;
	}
}

sub makemaker_args {
	my $self = shift;
	my $args = ( $self->{makemaker_args} ||= {} );
	%$args = ( %$args, @_ );
	return $args;
}

# For mm args that take multiple space-seperated args,
# append an argument to the current list.
sub makemaker_append {
	my $self = sShift;
	my $name = shift;
	my $args = $self->makemaker_args;
	$args->{name} = defined $args->{$name}
		? join( ' ', $args->{name}, @_ )
		: join( ' ', @_ );
}

sub build_subdirs {
	my $self    = shift;
	my $subdirs = $self->makemaker_args->{DIR} ||= [];
	for my $subdir (@_) {
		push @$subdirs, $subdir;
	}
}

sub clean_files {
	my $self  = shift;
	my $clean = $self->makemaker_args->{clean} ||= {};
	  %$clean = (
		%$clean,
		FILES => join ' ', grep { length $_ } ($clean->{FILES} || (), @_),
	);
}

sub realclean_files {
	my $self      = shift;
	my $realclean = $self->makemaker_args->{realclean} ||= {};
	  %$realclean = (
		%$realclean,
		FILES => join ' ', grep { length $_ } ($realclean->{FILES} || (), @_),
	);
}

sub libs {
	my $self = shift;
	my $libs = ref $_[0] ? shift : [ shift ];
	$self->makemaker_args( LIBS => $libs );
}

sub inc {
	my $self = shift;
	$self->makemaker_args( INC => shift );
}

my %test_dir = ();

sub _wanted_t {
	/\.t$/ and -f $_ and $test_dir{$File::Find::dir} = 1;
}

sub tests_recursive {
	my $self = shift;
	if ( $self->tests ) {
		die "tests_recursive will not work if tests are already defined";
	}
	my $dir = shift || 't';
	unless ( -d $dir ) {
		die "tests_recursive dir '$dir' does not exist";
	}
	%test_dir = ();
	require File::Find;
	File::Find::find( \&_wanted_t, $dir );
	$self->tests( join ' ', map { "$_/*.t" } sort keys %test_dir );
}

sub write {
	my $self = shift;
	die "&Makefile->write() takes no arguments\n" if @_;

	# Make sure we have a new enough
	require ExtUtils::MakeMaker;

	# MakeMaker can complain about module versions that include
	# an underscore, even though its own version may contain one!
	# Hence the funny regexp to get rid of it.  See RT #35800
	# for details.

	$self->configure_requires( 'ExtUtils::MakeMaker' => $ExtUtils::MakeMaker::VERSION =~ /^(\d+\.\d+)/ );

	# Generate the
	my $args = $self->makemaker_args;
	$args->{DISTNAME} = $self->name;
	$args->{NAME}     = $self->module_name || $self->name;
	$args->{VERSION}  = $self->version;
	$args->{NAME}     =~ s/-/::/g;
	if ( $self->tests ) {
		$args->{test} = { TESTS => $self->tests };
	}
	if ($] >= 5.005) {
		$args->{ABSTRACT} = $self->abstract;
		$args->{AUTHOR}   = $self->author;
	}
	if ( eval($ExtUtils::MakeMaker::VERSION) >= 6.10 ) {
		$args->{NO_META} = 1;
	}
	if ( eval($ExtUtils::MakeMaker::VERSION) > 6.17 and $self->sign ) {
		$args->{SIGN} = 1;
	}
	unless ( $self->is_admin ) {
		delete $args->{SIGN};
	}

	# merge both kinds of requires into prereq_pm
	my $prereq = ($args->{PREREQ_PM} ||= {});
	%$prereq = ( %$prereq,
		map { @$_ }
		map { @$_ }
		grep $_,
		($self->configure_requires, $self->build_requires, $self->requires)
	);

	# Remove any reference to perl, PREREQ_PM doesn't support it
	delete $args->{PREREQ_PM}->{perl};

	# merge both kinds of requires into prereq_pm
	my $subdirs = ($args->{DIR} ||= []);
	if ($self->bundles) {
		foreach my $bundle (@{ $self->bundles }) {
			my ($file, $dir) = @$bundle;
			push @$subdirs, $dir if -d $dir;
			delete $prereq->{$file};
		}
	}

	if ( my $perl_version = $self->perl_version ) {
		eval "use $perl_version; 1"
			or die "ERROR: perl: Version $] is installed, "
			. "but we need version >= $perl_version";
	}

	$args->{INSTALLDIRS} = $self->installdirs;

	my %args = map { ( $_ => $args->{$_} ) } grep {defined($args->{$_})} keys %$args;

	my $user_preop = delete $args{dist}->{PREOP};
	if (my $preop = $self->admin->preop($user_preop)) {
		foreach my $key ( keys %$preop ) {
			$args{dist}->{$key} = $preop->{$key};
		}
	}

	my $mm = ExtUtils::MakeMaker::WriteMakefile(%args);
	$self->fix_up_makefile($mm->{FIRST_MAKEFILE} || 'Makefile');
}

sub fix_up_makefile {
	my $self          = shift;
	my $makefile_name = shift;
	my $top_class     = ref($self->_top) || '';
	my $top_version   = $self->_top->VERSION || '';

	my $preamble = $self->preamble
		? "# Preamble by $top_class $top_version\n"
			. $self->preamble
		: '';
	my $postamble = "# Postamble by $top_class $top_version\n"
		. ($self->postamble || '');

	local *MAKEFILE;
	open MAKEFILE, "< $makefile_name" or die "fix_up_makefile: Couldn't open $makefile_name: $!";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;

	my @I_dirs = $self->_find_extra_inc;
	my $I_list = join(", ", map { qq{'$_'} } @I_dirs);
	my $I_flags = join(" ", map { qq{-I$_} } @I_dirs);
	my $I_flags_quoted = join(" ", map { qq{"-I$_"} } @I_dirs);
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1$I_list, /;
	# don't know why this isn't quoted but it used to be just -Ilib
	# so $I_flags does it the same way -- dagolden, 2009-04-07
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ $I_flags$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ $I_flags_quoted$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 $I_flags_quoted/m;
	$makefile =~ s/^(PERL = .*)/$1 $I_flags_quoted/m;

	# Module::Install will never be used to build the Core Perl
	# Sometimes PERL_LIB and PERL_ARCHLIB get written anyway, which breaks
	# PREFIX/PERL5LIB, and thus, install_share. Blank them if they exist
	$makefile =~ s/^PERL_LIB = .+/PERL_LIB =/m;
	#$makefile =~ s/^PERL_ARCHLIB = .+/PERL_ARCHLIB =/m;

	# Perl 5.005 mentions PERL_LIB explicitly, so we have to remove that as well.
	$makefile =~ s/(\"?)-I\$\(PERL_LIB\)\1//g;

	# XXX - This is currently unused; not sure if it breaks other MM-users
	# $makefile =~ s/^pm_to_blib\s+:\s+/pm_to_blib :: /mg;

	open  MAKEFILE, "> $makefile_name" or die "fix_up_makefile: Couldn't open $makefile_name: $!";
	print MAKEFILE  "$preamble$makefile$postamble" or die $!;
	close MAKEFILE  or die $!;

	1;
}

sub preamble {
	my ($self, $text) = @_;
	$self->{preamble} = $text . $self->{preamble} if defined $text;
	$self->{preamble};
}

sub postamble {
	my ($self, $text) = @_;
	$self->{postamble} ||= $self->admin->postamble;
	$self->{postamble} .= $text if defined $text;
	$self->{postamble}
}

sub _find_extra_inc {
	my ($self) = @_;
	# Adapted from Module::Build::Base
	my $quote = $^O eq 'MSWin32' ? q{"} : q{'};
	my @default_inc = qx/$^X -le $quote print for \@INC $quote/;
	chomp @default_inc;
	my %seen;
	$seen{$_}++ for @default_inc;
	return grep { !$seen{$_}++ } @INC;
}

1;

__END__

#line 396


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;



my %WriteMakefileArgs = (
  "ABSTRACT" => "sends test results to cpan-testers\@perl.org",
  "AUTHOR" => "Adam J. Foxson <afoxson\@pobox.com>, David Golden <dagolden\@cpan.org>, Kirrily \"Skud\" Robert <skud\@cpan.org>, Ricardo Signes <rjbs\@cpan.org>, Richard Soderberg <rsod\@cpan.org>, Kurt Starsinic <Kurt.Starsinic\@isinet.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Test-Reporter",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Test::Reporter",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Cwd" => 0,
    "File::Temp" => 0,
    "FileHandle" => 0,
    "Sys::Hostname" => 0,
    "Time::Local" => 0,
    "base" => 0,
    "constant" => 0,
    "strict" => 0,
    "vars" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "1.60",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Cwd" => 0,
  "Data::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Find" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => 0,
  "FileHandle" => 0,
  "List::Util" => 0,
  "Sys::Hostname" => 0,
  "Test::More" => "0.88",
  "Time::Local" => 0,
  "base" => 0,
  "constant" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



